keps/sig-api-machinery/2144-clientgo-apply/README.md

```diff
@@ -358,6 +284,61 @@ adding functions to interfaces is a change we have made in the past, and develop
 that have alternate implementations of the interface will usually get a compiler
 error in this case, which is relatively trivial to resolve
 
+### read/modify/write loop support
+
+While it is
+[recommended](https://kubernetes.io/docs/reference/using-api/server-side-apply/)
+that apply clients use "fully specified intent" rather than
+read/modify/write loops, there are many existing controllers that use
+a get/modify/update loop today, and are not easy to convert to the
+"fully specified intend" approach.
+
+For such controllers, providing a convenient way to do a
+"get-previously-applied/modify/apply" loop is pragmatic. It allows
+these controller to benefit from apply by sending a minimal apply
+patch (as opposed to sending the entire object for update) and to
+reduce the odds of conflict with other controllers.
+
+To support this, the builders will include "BuildApply" utility functions
+function. For example:
+
+```
+fieldManger := "my-field-manager"
+// 1. read
+deployment, err := client.AppsV1().Deployments(ns).Get("deployment-name", metav1.GetOptions{})
+if err != nil {
+  // handle err
+}
+applyConfig, err := appsv1apply.BuildDeploymentApply(deployment, fieldManager)
+if err != nil {
+  // handle err
+}
+
+// 2. modify
+applyConfig.GetSpec().Replicas(10)
+
+// 3. apply
+client.AppsV1().Deployments(ns).Apply(ctx, applyConfig, metav1.ApplyOptions{FieldManager: fieldManager, Force: true})
+```
+
+In the above example, `BuildDeploymentApply` constructs a populated
+apply configuration from a deployment object returned from a Get. It
+uses the provided field manager to get the matching field set
+(`FieldsV1` data) from object and then combines the field set with the
+object to produce the apply configuration.
+
+We will clearly document that on the "BuildApply" functions that we
+recommend using "fully specified intent" when using apply and also
+when it might be appropriate (and inappropriate) to use the "BuildApply"
+utility.
+
+Additionally, as a convenience, we may add a `GetForApply()` function
+to the generate client-go that returns an apply configuration directly
+(internally, it just calls Get and then BuildApply).
```

Comment:
I would have only made `GetForApply` available, so that people can't use `BuildApply` the wrong way (removes the need to protect people from doing `FromUnstructured` as described below).

Reply:
I was thinking about this earlier today, and I think I agree. There is no way to know if a developer has modified the go struct (some changes are possible to detect, but require checking all the field sets of all field managers, some are not possible to detect).

@lavalamp WDYT? Should we restrict the read/modify/write flow to be:

```
applied, err := client.GetApplied(..., "field-manager")
// modify in place
client.Apply(applied, ...)
```

?

Reply:
I didn't sleep well, and thought about it. The problem with tightly coupled things is that it's hard to re-use in a different context. It means that one can't create such an object from the result of anything but GET (create/update/patch).

Reply:
Another alternative would be to add a utility function that takes the client as an arg:

```
applied, err := appsv1apply.GetApplied(client, ctx, name, "field-manager", &metav1.GetOptions{})
```

But I prefer having the `GetApplied` function directly on the client.

Reply:
> I didn't sleep well, and thought about it. The problem with tightly coupled things is that it's hard to re-use in a different context. It means that one can't create such an object from the result of anything but GET (create/update/patch).

That's a really good point.

Reply:
I've updated the KEP and explained that having `client.GetForApply()` is a rejected alternative. We'll have to figure out what guard rails we can get in place to keep developers from modifying the object before converting it into an applied configuration.